### PR TITLE
Homework_analytical_patterns_sl2902

### PR DESCRIPTION
**Pull request by sl2902**

### CODE REVIEW

**Comment by julie-scherer** [comment]
## Automated Feedback from ChatGPT

### Feedback for `query_1.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_2.sql`

Overall, your code is well-structured and logically follows the given homework prompt. You have correctly used the `GROUPING SETS` clause to aggregate data along the required dimensions. However, there are a few areas where your code could be improved.

1) **Coverage of the homework prompt:** Your code does not fully cover the homework prompt. The prompt asks for aggregation along the dimensions of player and team, player and season, and team. However, in your `GROUPING SETS` clause, you have used `team_city` instead of `team_name`. Please ensure that you are using the correct column names as per the prompt.

2) **Readability:** Your code is generally readable, with meaningful variable names and clear formatting. However, you could improve the readability further by adding more comments to explain your thought process, especially in complex sections of the code like the `CASE` statement.

3) **Robustness and error handling:** Your code does not handle the case where `team_city` is null. In your `COALESCE` function, you replace null `team_city` values with 'N/A'. However, the example solution replaces null values with 'overall', which might be more meaningful in this context. Consider updating your `COALESCE` function to replace null values with 'overall' instead of 'N/A'.

4) **Modularity and reusability:** Your code is modular and the use of a CTE (Common Table Expression) makes it easy to understand and reuse specific components. However, you could improve the reusability of your code by creating a separate function for the `CASE` statement that calculates `total_games_won`. This would allow you to reuse this logic in other parts of your code if needed.

In conclusion, your code is well-structured and logically follows the homework prompt, but there are a few areas where it could be improved. Keep up the good work and continue to focus on improving your code's readability, robustness, and reusability.

### Feedback for `query_3.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_4.sql`

Overall, your solution is well-structured and logically follows the homework prompt. You've successfully written a query to find the player who scored the most points in a single season. However, there are a few areas where your solution could be improved.

1) **Coverage of the homework prompt:** Your solution does not use the `game_details_dashboard` table as specified in the prompt. Instead, you've used `sundeep.grouping_sets`. Please ensure you're using the correct table for your queries.

2) **Readability:** Your code is generally well-formatted and easy to read. You've used meaningful variable names and included comments to explain your thought process, which is excellent. However, the alias `n_r` for the rank is not very descriptive. Consider using a more descriptive alias like `rank`.

3) **Robustness and error handling:** Your code does not handle the case where `points` could be `NULL`. This could potentially lead to incorrect results. Consider adding a condition to filter out rows where `points` is `NULL`.

4) **Modularity and reusability:** Your use of a CTE (Common Table Expression) with the `WITH` clause is a good practice for modularity and reusability. It makes your code easier to read and understand.

Here's a revised version of your query considering the above points:

```sql
-- Find the player with the most
-- points earned in any season
WITH
    ranked AS (
        SELECT
            player_name,
            season,
            total_player_points,
            DENSE_RANK() OVER(ORDER BY total_player_points DESC) AS rank
        FROM
            game_details_dashboard
        WHERE
            player_name IS NOT NULL
        AND
            season IS NOT NULL
        AND
            total_player_points IS NOT NULL
)
SELECT
  player_name,
  season,
  total_player_points
FROM
  ranked
WHERE
  rank = 1
```

Keep up the good work, and continue to focus on the details of the assignment prompts.

### Feedback for `query_5.sql`

No feedback generated as no changes were detected in the current push.

**Comment by julie-scherer** [comment]
## Automated Feedback from ChatGPT

### Feedback for `query_1.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_2.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_3.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_4.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_5.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_6.sql`

Overall, your solution is well-structured and logically follows the given homework prompt. You have correctly used window functions to calculate the rolling sum of games won over a 90-game stretch for each team. However, there are a few areas where your solution could be improved.

1) **Coverage of the homework prompt:** Your solution does not fully cover the prompt. The prompt asks for the most games a single team has won in a given 90-game stretch, but your solution calculates the most games won in a 90-day stretch. This is a significant difference, as there may not be a game every day. You should adjust your window function to count games, not days. 

2) **Readability:** Your code is generally well-formatted and easy to read. You have used meaningful variable names and included comments to explain your thought process, which is good practice. However, the comment on the first line of your code is misleading, as it mentions a 90-day window instead of a 90-game window. Please correct this to avoid confusion.

3) **Robustness and error handling:** Your code does not handle potential errors or edge cases. For example, it does not account for the possibility of games ending in a tie or being cancelled. You could improve your code by adding error handling or checks for these situations.

4) **Modularity and reusability:** Your code is modular and well-structured, with separate CTEs for different parts of the calculation. This makes it easy to understand and reuse specific components in other contexts. However, you could improve the reusability of your code by defining the window size (90 games) as a variable at the start of your code, making it easy to adjust for different window sizes.

Here's how you could modify your code to count games instead of days:

```sql
WITH
 team_games AS ( 
   SELECT
    a.game_id,
    team_city,
    game_date_est,
    ROW_NUMBER() OVER(PARTITION BY team_city ORDER BY game_date_est) AS game_number,
    MAX(CASE WHEN (a.team_id = b.home_team_id AND b.home_team_wins = 1) OR
              (a.team_id = b.visitor_team_id AND b.home_team_wins = 0) THEN 1
    ELSE 0 END)
   AS games_won
   FROM
    bootcamp.nba_game_details a join bootcamp.nba_games b on a.game_id = b.game_id
   GROUP BY
     a.game_id,
     team_city,
     game_date_est
  
), streak AS (
  SELECT
   game_id,
   team_city,
   game_date_est,
   -- Calculate rolling sum over a 90 game window
   SUM(games_won) OVER(PARTITION BY team_city ORDER BY game_number
   ROWS BETWEEN 89 PRECEDING AND CURRENT ROW) AS rolling_games_won
 FROM
   team_games
)
SELECT
    team_city,
    MAX(rolling_games_won) AS streak_count
  FROM
   streak
 GROUP BY
   1
 ORDER BY
   2 DESC
```

This code adds a `game_number` column in the `team_games` CTE, which is then used in the window function in the `streak` CTE to count games instead of days.

### Feedback for `query_7.sql`

Overall, your solution is well-structured and logically follows the given homework prompt. You have successfully used window functions to answer the question: "How many games in a row did LeBron James score over 10 points a game?" However, there are a few areas where your code could be improved for readability, robustness, and reusability.

1) **Readability:** Your code is generally well-structured and easy to follow. However, it could benefit from more comments explaining your thought process, especially in the more complex parts of your code. For example, in the `lagged` and `streak_groups` CTEs, it would be helpful to explain why you're calculating `lag_diff` and `streak_group`, and how these contribute to the final result.

2) **Robustness and Error Handling:** Your code does not seem to handle potential errors or edge cases. For instance, what if the `player_name` does not exist in the database or there are no games where LeBron James scored over 10 points? It would be beneficial to add some error handling or checks to ensure the input data is valid.

3) **Modularity and Reusability:** Your code is modular and each CTE performs a specific task, which is good. However, the `player_name` and points threshold are hardcoded into your query. Consider parameterizing these values to make your code more reusable. For example, you could define variables at the beginning of your script for the player name and points threshold, and then use these variables in your query.

4) **Code Coverage:** Your code covers the main aspects of the homework prompt. However, it's important to note that your solution and the example solution take slightly different approaches to solve the problem. Both are valid, but the example solution uses a `ROW_NUMBER()` function to create a unique row number for each game and then subtracts the `streak_groups` to identify streaks. Your solution, on the other hand, uses the `LAG()` function to calculate the difference between the row numbers of consecutive games where LeBron James scored over 10 points. Both approaches are valid, but it's worth understanding the differences.

In conclusion, your solution is generally well-implemented, but could benefit from more comments, error handling, and parameterization of hardcoded values. Keep up the good work!
